pkg/operator/controller/gatewayclass/subscription.go

```diff
@@ -115,3 +121,76 @@ func subscriptionChanged(current, expected *operatorsv1alpha1.Subscription) (boo
 
 	return true, updated
 }
+
+// ensureServiceMeshOperatorInstallPlan attempts to ensure that the install plan for the appropriate OSSM operator
+// version is approved.
+func (r *reconciler) ensureServiceMeshOperatorInstallPlan(ctx context.Context) (bool, *operatorsv1alpha1.InstallPlan, error) {
+	haveInstallPlan, current, err := r.currentInstallPlan(ctx)
+	if err != nil {
+		return false, nil, err
+	}
+	switch {
+	case !haveInstallPlan:
+		// The OLM operator creates the initial InstallPlan, so if it doesn't exist yet, do nothing.
+		return false, nil, nil
+	case haveInstallPlan:
+		desired, err := desiredInstallPlan(current)
+		if err != nil {
+			return true, current, err
+		}
+		if updated, err := r.updateInstallPlan(ctx, current, desired); err != nil {
+			return true, current, err
+		} else if updated {
+			return r.currentInstallPlan(ctx)
+		}
+	}
+	return false, current, nil
+}
+
+func (r *reconciler) currentInstallPlan(ctx context.Context) (bool, *operatorsv1alpha1.InstallPlan, error) {
+	InstallPlans := &operatorsv1alpha1.InstallPlanList{}
+	if err := r.client.List(ctx, InstallPlans, client.InNamespace(serviceMeshOperatorNamespace)); err != nil {
+		return false, nil, err
+	}
+	if InstallPlans == nil || len(InstallPlans.Items) == 0 {
+		return false, nil, nil
+	}
+	for _, InstallPlan := range InstallPlans.Items {
+		for _, CSVName := range InstallPlan.Spec.ClusterServiceVersionNames {
+			if CSVName == r.config.GatewayAPIOperatorVersion {
+				return true, &InstallPlan, nil
+			}
+		}
+	}
+	return false, nil, fmt.Errorf("no InstallPlan with cluster service version %s found in namespace %s", r.config.GatewayAPIOperatorVersion, serviceMeshOperatorNamespace)
+}
+
+func desiredInstallPlan(current *operatorsv1alpha1.InstallPlan) (*operatorsv1alpha1.InstallPlan, error) {
```

Comment:
`desiredInstallPlan` would normally have all the fields set to what we want them to be.   It should not be a copy of the current install plan, even if you expect that will never change.

Reply:
I would generally agree, but I think this is a special case. The InstallPlan is really owned by the OLM operator, and the ingress operator is only managing the Approved field of this particular InstallPlan. Maybe there's a different way to write it or there's something I can write in the comments to explain why it deviates from the usual pattern, but do think this is the right logic for this particular case.